src/main/java/com/hotels/mutantswarm/MutantSwarmExtension.java

```diff
@@ -0,0 +1,212 @@
+      SwarmResultsBuilder swarmResultBuilder = new SwarmResultsBuilder(swarm,
+          context.getRequiredTestClass().toString());
+      contextRef.compareAndSet(null, new ExecutionContext(swarm, swarmResultBuilder));
+    }
+
+    mutants = contextRef.get().swarm.getMutants();
+    ExecutionContext executionContext = contextRef.get();
+    for (Mutant mutant : contextRef.get().swarm.getMutants()) {
+      MutatedSource mutatedSource = core.mutateSource(executionContext.getSource(), mutant);
+      mutatedSources.add(mutatedSource);
+    }
+
+    results = contextRef.get();
+
+    // Here we specify the number of times a test should be repeated, but the scripts will be mutated accordingly for
+    // each individual test in the postProcessTestInstance method
+    return IntStream.rangeClosed(1, mutants.size() + 1).mapToObj(repitition -> new MutantSwarmTestTemplate());
+  }
+
+  @Override
+  public void postProcessTestInstance(Object target, ExtensionContext extensionContext) {
+    testNumber++;
+
+    // The first test per swarm should run normally and without mutations
+    if (testNumber != 0) {
+      scriptsUnderTest.clear();
+      MutatedSource mutatedSource = mutatedSources.get(testNumber - 1);
+      List<MutantSwarmScript> MScripts = mutatedSource.getScripts();
+
+      for (Script s : MScripts) {
+        scriptsUnderTest.add(s);
+      }
+    }
+
+    super.postProcessTestInstance(target, extensionContext);
+
+  }
+
+  @Override
+  public void testSuccessful(ExtensionContext context) {
+    if (testNumber != 0) {
+      if (firstTestPassed == true) {
+        log.info("Mutant survived - bad");
+        results
+            .addTestOutcome(context.getDisplayName(), mutants.get(testNumber - 1),
+                mutatedSources.get(testNumber - 1).getMutation(), MutantState.SURVIVED);
+      }
+    } else {
+      firstTestPassed = true;
+    }
+  }
+
+  @Override
+  public void testFailed(ExtensionContext context, Throwable cause) {
+    if (testNumber != 0) {
+      if (firstTestPassed == true) {
+        log.info("Mutant killed - good. " + cause.toString());
+        results
+            .addTestOutcome(context.getRequiredTestMethod().toString(), mutants.get(testNumber - 1),
+                mutatedSources.get(testNumber - 1).getMutation(), MutantState.KILLED);
+      }
+    } else {
+      firstTestPassed = false;
+      // TODO: if the first test does not pass, then we should not run it again with the mutations because it's useless
+    }
+  }
+
+  @Override
+  public void afterAll(ExtensionContext context) throws Exception {
+    SwarmResults swarmResults = core.getSwarmResults(results);
+    log.debug("Finished testing. Generating report.");
+    new ReportGenerator(swarmResults).generate();
+  }
+
+  // This method sets the scripts for the first time to generate the swarm
+  private void setFirstScripts(ExtensionContext context) {
```

Comment:
I think we need to look at this in a bit more detail to see what we can change in HiveRunner so we don't have to repeat a lot of code here.